### PR TITLE
show validation errors on basic form inputs

### PR DESCRIPTION
**Pull request by briankariuki**
This commit adds form input validations, fixes issue #31, #32, #33

### CODE REVIEW

File: lib/animina/accounts/resources/user.ex
Comment (wintermeyer) [inline lib/animina/accounts/resources/user.ex:1]: Maybe a miscommunication from my side. The user should be able to use "doesn't matter". Please add that case as a valid possibility to the validations.

Bildschirmfoto 2024-02-15 um 07 51 41

File: lib/animina/validations/min_max_age.ex [hunk 1]
```diff
@@ -0,0 +1,35 @@
+defmodule Animina.Validations.MinMaxAge do
+  use Ash.Resource.Validation
+
+  @moduledoc """
+  This is a module for validating the mimimum and maximum partner ages
+  """
+
+  @impl true
+  def init(opts) do
+    if is_atom(opts[:attribute]) do
```
Comment (wintermeyer) [inline lib/animina/validations/min_max_age.ex]: In Elixir I prefer a `case` over an `if`. Especially in cases when there is a chance that more stuff gets added in the future. Please change this to a `case` statement.

BTW: An `if` becomes a `case` under the hood. https://hexdocs.pm/elixir/case-cond-and-if.html

Bildschirmfoto 2024-02-15 um 07 55 24

File: lib/animina/validations/min_max_height.ex [hunk 2]
```diff
@@ -0,0 +1,35 @@
+defmodule Animina.Validations.MinMaxHeight do
+  use Ash.Resource.Validation
+
+  @moduledoc """
+  This is a module for validating the mimimum and maximum partner heights
+  """
+
+  @impl true
+  def init(opts) do
+    if is_atom(opts[:attribute]) do
```
Comment (wintermeyer) [inline lib/animina/validations/min_max_height.ex]: Please use a `case` here.

File: lib/animina_web/live/root_live.ex [hunk 3]
```diff
@@ -71,21 +92,42 @@ defmodule AniminaWeb.RootLive do
...
+              f,
+              :username,
               class:
-                "block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6",
+                "block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset placeholder:text-gray-400 focus:ring-2 focus:ring-inset sm:text-sm  phx-no-feedback:ring-gray-300 phx-no-feedback:focus:ring-indigo-600 sm:leading-6 " <>
+                  if(get_field_errors(f[:username], :username) != [],
+                    do: "ring-red-600 focus:ring-red-600",
+                    else: "ring-gray-300 focus:ring-indigo-600"
+                  ),
               placeholder: "Pusteblume1977",
```
Comment (wintermeyer) [inline lib/animina_web/live/root_live.ex]: This should be a gettext("Pusteblume1977"). My bad. Please fix it while working here.

**Review by wintermeyer — CHANGES_REQUESTED** [review]
@briankariuki I like your code and the way you tackled the problem! Great work!

Requests:
- Please add screenshots to the PRs in case a GUI element is involved. So that I get a quick idea when first seeing the PR without having to pull the code.
- Please use one PR per page/problem. This is ok because it is one workflow but I would have splitted it up in two PRs.

File: lib/animina_web/live/root_live.ex [hunk 4]
```diff
@@ -71,21 +92,39 @@ defmodule AniminaWeb.RootLive do
...
           </label>
           <div class="mt-2">
-            <%= text_input(f, :username,
+            <%= text_input(
+              f,
+              :username,
               class:
-                "block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6",
+                "block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset  placeholder:text-gray-400 focus:ring-2 focus:ring-inset  sm:text-sm sm:leading-6 " <>
+                  if(get_field_errors(f[:username], :username) != [],
```
Comment (wintermeyer) [inline lib/animina_web/live/root_live.ex]: I haven't played with this but I think `if(get_field_errors(f[:username], :username) != []` can be done better with something like `unless(get_field_errors(f[:username], :username)`

```elixir
iex(1)> if [] do
...(1)> IO.puts "a"
...(1)> else
...(1)> IO.puts "b"
...(1)> end
a
:ok
``

File: lib/animina/accounts/resources/user.ex
Comment (briankariuki) [reply lib/animina/accounts/resources/user.ex:1]: The validation now works when the user selects the `doesn't matter` option.

image

File: lib/animina/validations/min_max_age.ex (hunk 1 above)
Comment (briankariuki) [reply lib/animina/validations/min_max_age.ex]: This has been replaced with case.

File: lib/animina_web/live/root_live.ex (hunk 4 above)
Comment (briankariuki) [reply lib/animina_web/live/root_live.ex]: I couldn't get unless to work like you suggested `unless(get_field_errors(f[:username], :username))` , instead I've opted to include `== []`  in the check.  `unless(get_field_errors(f[:username], :username) == [])`

**Review by wintermeyer — APPROVED** [review]

**Pull request merged**